File: 백준/Silver/24060. 알고리즘 수업 － 병합 정렬 1/알고리즘 수업 － 병합 정렬 1.cc
#include<stdio.h>
#define MAXSIZE 500000
int n,k,save;
int result;
int arr[MAXSIZE + 1];
int temp[MAXSIZE + 1];

void findResult(int num) {
	save++;
	if (save == k) {
		result = num;
	}
}
void merge(int start, int mid, int end) {
	int i = start;
	int j = mid + 1;
	int idx = start;

	while (i <= mid && j <= end) {
		if (arr[i] > arr[j]) {
			temp[idx] = arr[j];
			findResult(arr[j]);
			j++;
			idx++;
		}
		else {
			temp[idx] = arr[i];
			findResult(arr[i]);
			i++;
			idx++;
		}
	}

	if (i > mid) {
		while (j <= end) {
			temp[idx] = arr[j];
			findResult(arr[j]);
			idx++;
			j++;
		}
	}
	else {
		while (i <= mid) {
			temp[idx] = arr[i];
			findResult(arr[i]);
			idx++;
			i++;
		}
	}

	for (int i = start; i <= end; i++) {
		arr[i] = temp[i];
	}
}
void merge_sort(int start, int end) {
	if (start < end) {
		int mid = (start + end) / 2;
		merge_sort(start, mid);
		merge_sort(mid + 1, end);
		merge(start, mid, end);
	}
}
int main() {
	scanf("%d %d", &n, &k);

	for (int i = 0; i < n; i++) {
		scanf("%d", &arr[i]);
	}

	merge_sort(0, n-1);

	if (save < k) {
		printf("-1");
	}
	else {
		printf("%d", result);
	}
}

File: 백준/Silver/24060. 알고리즘 수업 － 병합 정렬 1/알고리즘 수업 － 병합 정렬 1.py
import sys

# 백준은 최대 재귀 깊이가 1000으로 설정되어 있어 변경 10^6으로
sys.setrecursionlimit(10**6)

def mergeSort(list,tmpList, lo, hi) :
    if lo >= hi : return
    mid = (lo + hi) // 2
    mergeSort(list,tmpList,lo,mid)
    mergeSort(list,tmpList,mid+1,hi)
    merge(list,tmpList,lo,mid,hi)

def merge(list,tmpList,lo,mid,hi):
    global cnt,result
    for i in range(lo,hi+1):
        tmpList[i] = list[i]
    
    i, j = lo,mid+1
    for idx in range(lo,hi+1):
        if i > mid : 
            list[idx], j = tmpList[j], j + 1
        elif j > hi :
            list[idx], i = tmpList[i], i + 1
        elif tmpList[j] <= tmpList[i] :
            list[idx], j = tmpList[j], j + 1
        else :
            list[idx], i = tmpList[i], i + 1
        cnt += 1
        if cnt == k:
                result = list[idx]

result = -1
cnt = 0
n, k = list(map(int, sys.stdin.readline().rsplit()))
unorederedList = list(map(int, sys.stdin.readline().rsplit()))
tmpList = [None for _ in range(n)]
mergeSort(unorederedList,tmpList,0,n - 1)
print(result)
